searx/engines/solidtorrents.py

```diff
@@ -1,51 +1,85 @@
 # SPDX-License-Identifier: AGPL-3.0-or-later
 # lint: pylint
-"""Solid Torrents
-
+"""SolidTorrents
 """
 
-from json import loads
+from datetime import datetime
 from urllib.parse import urlencode
+import random
+
+from lxml import html
+
+from searx.utils import extract_text, eval_xpath, eval_xpath_getindex
 
 about = {
     "website": 'https://www.solidtorrents.net/',
     "wikidata_id": None,
     "official_api_documentation": None,
-    "use_official_api": True,
+    "use_official_api": False,
     "require_api_key": False,
-    "results": 'JSON',
+    "results": 'HTML',
 }
 
 categories = ['files']
 paging = True
 
-base_url = 'https://www.solidtorrents.net/'
-search_url = base_url + 'api/v1/search?{query}'
+base_url = ''
+base_url_rand = ''
+
+units = {"B": 1, "KB": 2 ** 10, "MB": 2 ** 20, "GB": 2 ** 30, "TB": 2 ** 40}
+
+
+def size2int(size_str):
+    n, u = size_str.split()
+    return int(float(n.strip()) * units[u.strip()])
 
 
 def request(query, params):
-    skip = (params['pageno'] - 1) * 20
-    query = urlencode({'q': query, 'skip': skip})
+    global base_url_rand  # pylint: disable=global-statement
+    if isinstance(base_url, list):
+        base_url_rand = random.choice(base_url)
+    else:
+        base_url_rand = base_url
+    search_url = base_url_rand + '/search?{query}'
+    page = (params['pageno'] - 1) * 20
+    query = urlencode({'q': query, 'page': page})
     params['url'] = search_url.format(query=query)
-    logger.debug("query_url --> %s", params['url'])
     return params
 
 
 def response(resp):
     results = []
-    search_results = loads(resp.text)
-
-    for result in search_results["results"]:
-        results.append(
-            {
-                'infohash': result["infohash"],
-                'seed': result["swarm"]["seeders"],
-                'leech': result["swarm"]["leechers"],
-                'title': result["title"],
-                'url': "https://solidtorrents.net/view/" + result["_id"],
-                'filesize': result["size"],
-                'magnetlink': result["magnet"],
-                'template': "torrent.html",
-            }
-        )
+    dom = html.fromstring(resp.text)
+
+    for result in eval_xpath(dom, '//div[contains(@class, "search-result")]'):
+        a = eval_xpath_getindex(result, './div/h5/a', 0, None)
+        if a is None:
+            continue
+        title = extract_text(a)
+        url = eval_xpath_getindex(a, '@href', 0, None)
+        stats = eval_xpath(result, './div//div[contains(@class, "stats")]/div')
+        filesize = size2int(extract_text(stats[1]))
+        leech = extract_text(stats[2])
+        seed = extract_text(stats[3])
+        magnet = eval_xpath_getindex(result, './div//a[contains(@class, "dl-magnet")]/@href', 0, None)
+
+        params = {
+            'seed': seed,
+            'leech': leech,
+            'title': title,
+            'url': base_url_rand + url,
```

Comment:
> To keep exactly the same value `base_url` must be stored in `params` and then retrieve using `resp.search_params["base_url"]`.

Hm, I AFAIK request & response are in the same process environment .. isn't it?

I copied this method from

https://github.com/searxng/searxng/blob/de32d543bc5e7fac968a11c7cdd0e3b8f349a94d/searx/engines/invidious.py#L43-L49

https://github.com/searxng/searxng/blob/de32d543bc5e7fac968a11c7cdd0e3b8f349a94d/searx/engines/invidious.py#L67-L73

Reply:
Two different threads ( = two different user queries) can call the `request` function in a row and then the `response` function. The namespace will be same since this is the same engine.

Reply:
:facepalm: thanks for clarify, I fixed it here in solidtorrents engine ind in invidious engine https://github.com/searxng/searxng/pull/867